tests/expected/function-contract/valid_ptr.rs

```diff
@@ -0,0 +1,61 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// kani-flags: -Zfunction-contracts -Zmem-predicates
+
+//! Test that it is sound to use `assert_valid_ptr` inside a contract pre-condition.
+//! We cannot validate post-condition yet. This can be done once we fix:
+//! <https://github.com/model-checking/kani/issues/2997>
+#![feature(pointer_is_aligned)]
+
+mod pre_condition {
+    /// This contract should succeed only if the input is a valid pointer.
+    #[kani::requires(kani::mem::assert_valid_ptr(ptr) && ptr.is_aligned())]
```

Comment:
I find it ugly/non-intuitive to use an assertion inside a requires/ensures. I would prefer if we instead provide a separate attribute for pointer validity whose semantics are unambiguous. For example, we could add `#[kani::assert_valid_ptr(...)]` to make it clear that this attribute only asserts, both in the context of a `proof_for_contract` and `stub_verified`.

Reply:
I totally agree, but I'm also optimistic that this is a temporary solution until we have a proper method to query pointer validity. I'm not sure it grants a new attribute.

In fact, we would need two different attributes, one for pre and another for post-condition. We would also need to define order of evaluation, which might not be straight forward. This would also not work well if the pointer validity check is only relevant if another predicate is true.